### ISSUE TITLE
Url scheme’s 

### ISSUE THREAD

**Issue by gedw99** [gioui-plugins/gio-plugins#27]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

associate a file type with my gio app. So that when the user double clicks the file, it opens my gio app.

**Describe alternatives you've considered**

none.

**Additional context**

seen it also work for web URLs on desktop and mobile … 

https://stackoverflow.com/questions/67400225/handling-url-schemes-with-golang

example code for max and desktop

https://github.com/scraly/learning-go-by-examples/issues/7

**Comment by gedw99** [comment]
Good explanations 

https://developer.shotgridsoftware.com/af0c94ce/

**Comment by gedw99** [comment]
This is interesting aspect too 

https://help.twixlmedia.com/hc/en-us/articles/115000739265-Using-custom-URL-Schemes-in-your-app

Using the url schemes to trigger things in the app or it seems to deep link into sone part of your app.

**Comment by inkeliz** [comment]
The feature you mentioned sounds nice. However, implementing it will likely require modifying the gio and gigio libraries, which cannot be done on a third-party library. Additionally, changes to the manifest or plist files may be necessary on iOS and Android. This means that initially, only basic support for Linux and Windows as a "plugin" may be possible.

However, since we are changing gio/gogio: it may be possible to handle "deep links" (which are URLs that support HTTP/HTTPS) and are similar in nature. It might be beneficial to implement both features (custom schemes and deep links) together.

**Comment by gedw99** [comment]
thanks @inkeliz for your expertise on this.

Yep needs work on he GIo side, so I raised an issue on the GIO list at https://todo.sr.ht/~eliasnaur/gio/470 to get the ball rolling.   

I agree that Deep Links and File handlers ( if that's a good name for it ) have some crossover and it would be awesome to have both in GIO.

**Comment by gedw99** [comment]
https://github.com/fyne-io/fyne/issues/1847 has good details on how to make this work on mac, linux and wndows.

**Comment by gedw99** [comment]
Hey @inkeliz 

i put together a darwin / ios sample of this. I was wondering if i can PR it on a branch ?

**Comment by gedw99** [comment]
also because there is some forking of gio should we establish that fork somewhere so that all code can use it ?

**Comment by inkeliz** [comment]
I think is fine to PR. That will also make easier to understand what kind of changes need to apply to Gio. 👍

**Comment by gedw99** [comment]
I think we can make it work for browsers too.

so when you put it into the browser it opens the local app, or the web site if the local app is not installed.

we need proper packaging for mac and windows for this to work too.
I git it for mac and am working it for windows msi.

**Comment by inkeliz** [comment]
That feature is a "pre-requisite" to implement "OAuth" (SignIn with {Google/Microsoft/Apple/Facebook/WeChat}). Exists two strategies that can be applied:

- Localhost: On desktop create a localhost and then use `redirect_url` to `https://localhost`.
- Custom Scheme: On mobile/desktop listen to specific scheme (`your.app.something://`) and use it as `redirect_url`.

_In both cases PKCE is recommended for security._

_Exists another strategy, which is the app listen the some custom server (non-Oauth) and execute the login as Web, then the server (non-OAuth server) notifies the user. But that requires more setup._

-----

Since I'm working with `Login with ______`, then "Custom Scheme" will be added too. HOWEVER, that requires modify `gogio`, which leads again to "create a custom gogio".

**Comment by gedw99** [comment]
Hey @inkeliz 

sorry i got stuck on the PR, and side tracked into other work. 

The plan you have sounds awesome.  If i can help with any of it please tell me, as i have some time at the moment.

Also you mentioned the GOGIO cmd needing work; I have to modify GOGIO anyway in order to get proper web manifest for Web push notifications.
I wrote it up here: https://github.com/binwiederhier/ntfy/issues/797
Is this aligned with what we need for Url Scheme, deep link ? i think it is but you tell me..

**Comment by gedw99** [comment]
Oh and i am planning to add Hankp Passkey support. 
Once we have GOGIO doing proper web manifest it will be pretty easy.

https://github.com/teamhanko/hanko/tree/main/quickstart has everything needed.
We jus need to hook gio gui into it. Or use Webview.

**Comment by inkeliz** [comment]
I'm currently (kind of) designing the API. The major issues, currently is:

- Registering on Windows is at runtime. But, on MacOS, iOS, Android it's at "compiled time" (on manifest ~gogio issue aside).
- Windows don't have any specific "Event". So, if you click at `mycustomproto://something`, it will launch another instance of the app (say, open another "myapp.exe"). It's possible to fix that using IPC (unix socket and similar). But... That only works if only one instance of the app is open. Imagine that you have 2x `myapp.exe` opened (it opens without any scheme, just by double-clicking at the icon). What should happen when you click at `myapp://do-somehting`?! Both must get notified, one of them or must be impossible to have 2 instances of the same program?

The implementation for Android is done (but, requires to modify the manifest manually, on gogio soruce-code, currently). The implementation for MacOS is almost done, and iOS should work by luck, I didn't test. The Windows is the next, but have that "undefined behaviour" of multiple same app opened.

**Comment by gedw99** [comment]
@inkeliz 

about windows opening the same app twice… I don’t have windows so can’t confirm but assuming your right then i aspect we let it happen ( if that how windows apps operate , or we use a config file to track open files of the file extension. If a 2nd attempts to open the same file we close and “ redirect “ to the already opened one.

This is just a brain storm idea as an alternative to ipc.

**Comment by gedw99** [comment]
I plan to put up a fork with lots of automation / ci for gio-plugins. So if you like it it can be PRed to your repo.

will let you know when that’s ready and you can have a look

**Comment by inkeliz** [comment]
https://github.com/gioui-plugins/gio-plugins/commit/df63942816ca053fdd649137c1f9dd555d74db39 that change adds support for Android/Windows. But, iOS/macOS needs to be upstream (Gio-core). So, I'm moving everything to Gio-core, but the Windows situation is quite tricky. Because even using IPC it needs to be done before Gio inits, otherwise it will show display some random UI and then close. Let's see what Elias will suggest.

You can PR anytime. :)

**Comment by inkeliz** [comment]
On Android, gogio must be patched:

```diff
diff --git a/gogio/androidbuild.go b/gogio/androidbuild.go
index cddba99..72d1d04 100644
--- a/gogio/androidbuild.go
+++ b/gogio/androidbuild.go
@@ -462,12 +462,24 @@ func exeAndroid(tmpDir string, tools *androidTools, bi *buildInfo, extraJars, pe
 			android:theme="@style/Theme.GioApp"
 			android:configChanges="screenSize|screenLayout|smallestScreenSize|orientation|keyboardHidden"
 			android:windowSoftInputMode="adjustResize"
+			android:launchMode="singleTask"
 			android:exported="true">
 			<intent-filter>
 				<action android:name="android.intent.action.MAIN" />
 				<category android:name="android.intent.category.LAUNCHER" />
 			</intent-filter>
 		</activity>
+		<activity android:name="com.inkeliz.deeplink.deeplink_android" 
+			android:label="com.inkeliz.deeplink.deeplink_android"
+			android:noHistory="true" 
+			android:launchMode="singleInstance">
+			<intent-filter>
+				<action android:name="android.intent.action.VIEW"></action>
+				<category android:name="android.intent.category.DEFAULT"></category>
+				<category android:name="android.intent.category.BROWSABLE"></category>
+				<data android:scheme="yourscheme"></data>
+			</intent-filter>
+		</activity>
 	</application>
 </manifest>`)
 	var manifestBuffer bytes.Buffer
```

**Comment by gedw99** [comment]
This looks really good. I will try out the  the deep links code and PR anything i see that can help.

Also since we have to change some things in  to gogio, could you let me know where your fork of gioio repo is ?
So then  i can use it when working on gio-plugins.

thanks

**Comment by inkeliz** [comment]
I didn't fork it, because `<data android:scheme="yourscheme"></data>` needs to change per-project. So, the best way is to create one `-deeplink` flag in gogio.

Since I'm moving it to Gio, you can track it here: https://github.com/gioui/gio-cmd/pull/9 (and another PR is coming to gioui/gio)

**Comment by gedw99** [comment]
> Gio



> I didn't fork it, because `<data android:scheme="yourscheme"></data>` needs to change per-project. So, the best way is to create one `-deeplink` flag in gogio.
> 
> Since I'm moving it to Gio, you can track it here: [gioui/gio-cmd#9](https://github.com/gioui/gio-cmd/pull/9) (and another PR is coming to gioui/gio)

@inkeliz 

Ok make sense now - thanks for the update.

**Comment by inkeliz** [comment]
While testing iOS, I notice one "bug": https://stackoverflow.com/questions/42839716/ios-tel-and-mailto-links-only-work-with-target-blank

For some reason, `<a href="yourapp://data"></a>` will not work, but adding `target="_blank"` fixes the issue, but I don't know why.

**Comment by gedw99** [comment]
Curious what happens when the user just types it into safari omni bar and it tries to open the installed app that matches the url scheme …

**Comment by inkeliz** [comment]
If you type into Safari's address bar it works fine, just the HTML (without _blank) doesn't work. 


I push this feature to Gio:
https://github.com/gioui/gio-cmd/pull/9
https://github.com/gioui/gio/pull/117

Maybe will take some couple of weeks to get it into Gio-core, _if that get's accepted_.

**Comment by gedw99** [comment]
I will try it out off the githash on your PR’s. 

This is really cool @inkeliz
